Bugfix/ts recent

rma-rripken opened the pull request:
Changes to allow testing related to
https://github.com/USACE/cwms-data-api/issues/604

rma-rripken committed:
Recent change added roles to ts_recent when its just a get and shouldn't need them

rma-rripken committed:
Was previously selecting tsView.TS_ID twice.  Made category and group optional.

rma-rripken committed:
Allow user to specify group or category.  Added detail to doc about tsIds needing to be comma separated.

rma-rripken committed:
Minor cleanup

rma-rripken committed:
Minor formatting

rma-rripken committed:
Minor formatting

rma-rripken committed:
Recent should not take group-id

rma-rripken committed:
Remove tsView CWMS_TS_ID from select

rma-rripken committed:
group should be a query param. not a path param

rma-rripken committed:
Ambiguous column happening in the version that takes tsIds too

MikeNeilson reviewed (changes requested):
Okay, so it works. but it sends back *every* possible unit.

MikeNeilson commented:
The result of testing:
```
[
    {
        "id": "Black Butte.Stor-Top Con.Inst.~1Day.0.Calc-val",
        "dqu": {
            "id": {
                "office-id": "SPK",
                "ts-code": 119359044,
                "unit-id": "1000 m3",
                "date-time": 1712559600000
            },
            "cwms-ts-id": "Black Butte.Stor-Top Con.Inst.~1Day.0.Calc-val",
            "version-date": -27079747200000,
            "value": -0.0123348183754752,
            "quality-code": 0,
            "start-date": 1704067200000,
            "end-date": 1735689600000
        }
    },
    {
        "id": "Black Butte.Stor-Top Con.Inst.~1Day.0.Calc-val",
        "dqu": {
            "id": {
                "office-id": "SPK",
                "ts-code": 119359044,
                "unit-id": "ac-ft",
                "date-time": 1712559600000
            },
            "cwms-ts-id": "Black Butte.Stor-Top Con.Inst.~1Day.0.Calc-val",
            "version-date": -27079747200000,
            "value": -0.01,
            "quality-code": 0,
            "start-date": 1704067200000,
            "end-date": 1735689600000
        }
    },
    {
        "id": "Black Butte.Stor-Top Con.Inst.~1Day.0.Calc-val",
        "dqu": {
            "id": {
                "office-id": "SPK",
                "ts-code": 119359044,
                "unit-id": "dsf",
                "date-time": 1712559600000
            },
            "cwms-ts-id": "Black Butte.Stor-Top Con.Inst.~1Day.0.Calc-val",
            "version-date": -27079747200000,
            "value": -0.005041666666666667,
            "quality-code": 0,
            "start-date": 1704067200000,
            "end-date": 1735689600000
        }
    },
    {
        "id": "Black Butte.Stor-Top Con.Inst.~1Day.0.Calc-val",
        "dqu": {
            "id": {
                "office-id": "SPK",
                "ts-code": 119359044,
                "unit-id": "ft3",
                "date-time": 1712559600000
            },
            "cwms-ts-id": "Black Butte.Stor-Top Con.Inst.~1Day.0.Calc-val",
            "version-date": -27079747200000,
            "value": -435.6,
            "quality-code": 0,
            "start-date": 1704067200000,
            "end-date": 1735689600000
        }
    },
    {
        "id": "Black Butte.Stor-Top Con.Inst.~1Day.0.Calc-val",
        "dqu": {
            "id": {
                "office-id": "SPK",
                "ts-code": 119359044,
                "unit-id": "gal",
                "date-time": 1712559600000
            },
            "cwms-ts-id": "Black Butte.Stor-Top Con.Inst.~1Day.0.Calc-val",
            "version-date": -27079747200000,
            "value": -3258.5140997796807,
            "quality-code": 0,
            "start-date": 1704067200000,
            "end-date": 1735689600000
        }
    },
    {
        "id": "Black Butte.Stor-Top Con.Inst.~1Day.0.Calc-val",
        "dqu": {
            "id": {
                "office-id": "SPK",
                "ts-code": 119359044,
                "unit-id": "kaf",
                "date-time": 1712559600000
            },
            "cwms-ts-id": "Black Butte.Stor-Top Con.Inst.~1Day.0.Calc-val",
            "version-date": -27079747200000,
            "value": -0.000009999999999999999,
            "quality-code": 0,
            "start-date": 1704067200000,
            "end-date": 1735689600000
        }
    },
    {
        "id": "Black Butte.Stor-Top Con.Inst.~1Day.0.Calc-val",
        "dqu": {
            "id": {
                "office-id": "SPK",
                "ts-code": 119359044,
                "unit-id": "mile3",
                "date-time": 1712559600000
            },
            "cwms-ts-id": "Black Butte.Stor-Top Con.Inst.~1Day.0.Calc-val",
            "version-date": -27079747200000,
            "value": -2.959280303030303e-9,
            "quality-code": 0,
            "start-date": 1704067200000,
            "end-date": 1735689600000
        }
    },
    {
        "id": "Black Butte.Stor-Top Con.Inst.~1Day.0.Calc-val",
        "dqu": {
            "id": {
                "office-id": "SPK",
                "ts-code": 119359044,
                "unit-id": "kgal",
                "date-time": 1712559600000
            },
            "cwms-ts-id": "Black Butte.Stor-Top Con.Inst.~1Day.0.Calc-val",
            "version-date": -27079747200000,
            "value": -3.258514099779681,
            "quality-code": 0,
            "start-date": 1704067200000,
            "end-date": 1735689600000
        }
    },
    {
        "id": "Black Butte.Stor-Top Con.Inst.~1Day.0.Calc-val",
        "dqu": {
            "id": {
                "office-id": "SPK",
                "ts-code": 119359044,
                "unit-id": "km3",
                "date-time": 1712559600000
            },
            "cwms-ts-id": "Black Butte.Stor-Top Con.Inst.~1Day.0.Calc-val",
            "version-date": -27079747200000,
            "value": -1.2334818375475201e-8,
            "quality-code": 0,
            "start-date": 1704067200000,
            "end-date": 1735689600000
        }
    },
    {
        "id": "Black Butte.Stor-Top Con.Inst.~1Day.0.Calc-val",
        "dqu": {
            "id": {
                "office-id": "SPK",
                "ts-code": 119359044,
                "unit-id": "m3",
                "date-time": 1712559600000
            },
            "cwms-ts-id": "Black Butte.Stor-Top Con.Inst.~1Day.0.Calc-val",
            "version-date": -27079747200000,
            "value": -12.3348183754752,
            "quality-code": 0,
            "start-date": 1704067200000,
            "end-date": 1735689600000
        }
    },
    {
        "id": "Black Butte.Stor-Top Con.Inst.~1Day.0.Calc-val",
        "dqu": {
            "id": {
                "office-id": "SPK",
                "ts-code": 119359044,
                "unit-id": "mcm",
                "date-time": 1712559600000
            },
            "cwms-ts-id": "Black Butte.Stor-Top Con.Inst.~1Day.0.Calc-val",
            "version-date": -27079747200000,
            "value": -0.0000123348183754752,
            "quality-code": 0,
            "start-date": 1704067200000,
            "end-date": 1735689600000
        }
    },
    {
        "id": "Black Butte.Stor-Top Con.Inst.~1Day.0.Calc-val",
        "dqu": {
            "id": {
                "office-id": "SPK",
                "ts-code": 119359044,
                "unit-id": "mgal",
                "date-time": 1712559600000
            },
            "cwms-ts-id": "Black Butte.Stor-Top Con.Inst.~1Day.0.Calc-val",
            "version-date": -27079747200000,
            "value": -0.003258514099779681,
            "quality-code": 0,
            "start-date": 1704067200000,
            "end-date": 1735689600000
        }
    },
    {
        "id": "Black Butte.Stor-Top Con.Inst.~1Day.0.Calc-val",
        "dqu": {
            "id": {
                "office-id": "SPK",
                "ts-code": 119359044,
                "unit-id": "kdsf",
                "date-time": 1712559600000
            },
            "cwms-ts-id": "Black Butte.Stor-Top Con.Inst.~1Day.0.Calc-val",
            "version-date": -27079747200000,
            "value": -0.000005041666666666667,
            "quality-code": 0,
            "start-date": 1704067200000,
            "end-date": 1735689600000
        }
    }
]
```

Pretty sure we don't want it to do that.

rma-rripken commented:
I think it fixes the bug.   We probably want the user to specify the unit, or the unit system and return the default for that unit sytsem. 
Also, we probably don't want to expose the ts-code.

rma-rripken commented:
This end-point was this way b/c that is what cwms-mobile was doing.   In cwms-mobile I believe it stored the units to use for each ts assignment and then selected the right entry from these results.  Not saying thats perfect or what we want to do - just explaining that this code is the way it is b/c it started from cwms-mobile.

MikeNeilson commented:
That certainly makes sense for why it's doing that but we probably shouldn't do that going forward. But to confirm *this* is the previously existing behavior and thus we're just resetting to it at least working?

rma-rripken committed:
Merge branch 'develop' into bugfix/ts_recent

rma-rripken committed:
Removed TsvDquId class, Removed TS_CODE.  Added immutable style builder to TsvDqu. Added dto TsvDquTest.

rma-rripken committed:
Merge branch 'develop' into bugfix/ts_recent

rma-rripken committed:
Reworked queries to add default units and to use aliases instead of looking up fields by name.
TRIM should be tagged as Boolean
Added UnitSystem argument
Added UnitSystem enum to Controllers

MikeNeilson commented on cwms-data-api/src/main/java/cwms/cda/ApiServlet.java:
```diff
@@ -635,7 +637,7 @@ private static String getAccessManagerName() {
     @Override
     protected void service(HttpServletRequest req, HttpServletResponse resp)
-            throws ServletException, IOException {
+            throws IOException {
```
Isn't the `throws ServletException` part of the interface definition?

rma-rripken replied:
Yes.  Sub-classes are allowed to narrow the list.  I put it back.

MikeNeilson reviewed (commented):
One question about a removed throws element.

rma-rripken committed:
Restore ServletException
